### PR TITLE
Add Trajectory Planner to AI

### LINKED ISSUE #2496: Update LinearVelocityAgent to stop at destination point
### Description of the task
`LinearVelocityAgent` represents an Agent in HRVO which only moves to its destination and does not consider collision with other Agents. This Agent is used to represent enemy robots, and the destination point is picked to be the edge of the field, given their current velocity. However, currently the `LinearVelocityAgent` continues moving past its destination without stopping, which is not realistic. 
A good starting point is `LinearVelocityAgent::computeNewVelocity()` and understanding how `HRVOSimulator::doStep()` and `Agent::update()` work.

### Acceptance criteria
- [ ] Update `LinearVelocityAgent` such that it stops at the desired destination point
- [ ] Test results using existing and new tests in `hrvo_test.cpp`

### LINKED ISSUE #2519: Re-enable DISABLED tests due to HRVO integration
### Description of the task
A set of `SimulatedErForceSimTacticTestFixture` tests have been DISABLED due to the replacement of the trajectory planner with HRVO. Furthermore, during the initial integration, ThetaStar path planner was still being used. Given the method which ThetaStar uses when the start of the path is in a obstacle, the robot can oscillate back and forth, and never reach its destination. Once more of the `New Motion Planner` project has been integrated, many of these tests should be able to enabled again. 
With the update implementation, magic numbers in `hrvo_agent.h` (`decel_dist_multiplier` `decel_pref_speed_multiplier`) should also be removed.

### Acceptance criteria
Enable (by removing `DISABLED_` from the start of the test name, or uncommenting the test) all disabled tests:
- [ ] enable `PenaltyKickEnemyPlayTest`
- [ ] enable `ChipTacticTest`
- [ ] enable `KickTacticTest`
- [ ] enable `MoveGoalieToGoalLineTacticTest`
- [ ] enable `ShadowEnemyTacticTest`
- [x] enable `SimulatedHRVOTest.test_zig_zag_movement`
- [ ] enable `CreaseDefenderEnvironment`
- [ ] Remove magic number

### LINKED ISSUE #2629: proj_path_planner: Make HRVO change the destination endpoint when it is blocked by a dynamic obstacle
### Description of the task

HRVO sometimes collides with opposing robots if a path point is blocked by a robot. We want to move this path point if we see that it is blocked by any obstacle (static or dynamic)

`Simulator::updateWorld,` `Simulator::updatePrimitiveSet` or `HRVOAgent::computeNewVelocity` are all good spots to consider when changing HRVO's destination. `EnlsvgPathPlanner::findClosestUnblockedCell `could be a model to find an unblocked grid cell

### Acceptance criteria

- [ ] Write a simulated_hrvo_test that sets up a situation where the destination is inside an obstacle
- [ ] Make changes to HRVO to move the endpoint if it is blocked

### LINKED ISSUE #2655: Thunderscope: Add/Remove HRVOLayers dynamically based on number of robots
### Description of the task
Currently we add exactly 6 HRVO layers to Thunderscope, however, we would like to improve this by dynamically adding and removing HRVO layers as robots gets added or removed from the field.
To implement this feature, we could add a new function for removing a layer, similar to the `Field.add_layer()` function, but using [removeItem](https://pyqtgraph.readthedocs.io/en/latest/graphicsItems/plotitem.html#pyqtgraph.PlotItem.removeItem).

### Acceptance criteria
- [ ] Dynamically add/remove HRVO layers based on number of robots
- [ ] Update Field legend accordingly
- [ ] Removed all TODO (\#2655) comments

### LINKED ISSUE #2676: HRVO: Dependency inject RobotNavigationObstacleConfig to HRVO
### Description of the task
HRVOAgent uses `RobotNavigationObstacleConfig`, however, right now its not dependency injected and is not updated on value changes, so if any values/parameters are changed through Thunderscope, it will not be reflected in HRVO.

### Acceptance criteria
- [ ] Dependency inject the dynamic parameters
- [ ] Move all constants in the HRVO directory which we might want to dynamically tune to a new dynamic parameter.

### Blocked By
#2666

### LINKED ISSUE #2894: Support different HRVO configurations and responsiveness
### Description of the task
We can make our robots more responsive by increasing their k_p in `HRVOConfig` which is used in the proportional controller in `HRVOAgent::computePreferredVelocity`. Other constants such max accel/decel are other effective ways to make a robot more responsive (they have to be tuned in caution to avoid ramping the motors to fast and breaking them). This is specifically useful for robots that need to be more responsive such as the goalie. 
One approach to supporting this would be to have a field in the primitive which has the specification for which config should be used.

Preliminary tests showed that k_p = 3.0 seems to work well for the goalie, with a balance of not overshooting in short distances while being more responsive (refer to "Simulation & IRL Motion Planning Synchronization" in Motion Planning Notion). Note that higher k_p results in overshooting when robot has to travel far distances. It is recommended that this is tested further.

### Acceptance criteria

- [ ] Determine the optimal k_p for the goalie
- [ ] Support robots dynamically changing their `HRVOConfig`

### Blocked By
PID HRVO PR (#2896)

### LINKED ISSUE #2913: ENLSVG can not find path from above enemy defense area to below
### Description of the task
As described in title. An example of this can be found in `TestEnlsvgPathPlanner.test_going_from_above_enemy_defense_area_to_below` which has the following starting params
```cpp
    // Above defense area
    Point start{4, 2};
    // Below defense area
    Point dest{4, -2};

    // Obstacle: INFLATED_ENEMY_DEFENSE_AREA
```

### Acceptance criteria

- [ ] Fix bug and enable test

### PR DESCRIPTION
**Pull request by nimazareian**
### Description
This PR adds support for motion planning and obstacle avoidance using a trajectory planner. This updated our motion planning architecture from running ENLSVG/GlobalPathPlanner in fullsystem and HRVO on the robots to running a multi-purpose trajectory planner in AI. 

This new architecture aims to unify the motion planning stack, while making it more maintainable and extensible in the long term. Furthermore, by generating trajectories all the way to the destination, we are able to make smarter choices in AI.

It is worth noting that although this is a large PR, there are various features that it are missing, most of which are tracked in various GH issues (e.g. #3104). As such, this is not a perfect finalized solution. Any suggestions for improvements are encouraged and welcomed! :smiley:  

The sampling of trajectory paths are based on [Tigers Mannheim's 2019 TDP](https://download.tigers-mannheim.de/release2019/ETDP_TIGERs_Mannheim_2019.pdf). 

https://github.com/UBC-Thunderbots/Software/assets/28585597/2df656df-af79-447e-b935-846c499ab2b3

### Core New Functionality & Review Checklist
- `trajectory_planner.h/cpp`: Implementation of the trajectory planner and cost functions used to get the best trajectory from a sample of trajectories.
- `primitive.h` + `move_primitive.h/cpp` + `stop_primitive.h/cpp`: New classes which are generated by tactics to set their intent. The `MovePrimitive` class first estimates the cost of primitive (duration of a direct trajectory to the destination, ignoring obstacles), and if the tactic as assigned to a robot, the `MovePrimitive` class will call the trajectory planner to generate an actual trajectory which avoids obstacles.
- `play.h/cpp` + `assigned_tactics_play.h/cpp`: Handles the `Primitive`s returned by tactic
- `primitive_exectory.h/cpp` + `network.h/cpp`: Handle new `TbotsProto::Primitive` msgs which includes the set of parameters used to generate the linear and angular trajectories in AI. This allows us to send minimum data from AI to robot to recreate the trajectory. Field testing showed that using velocity feedback is helpful as it is much more instantaneous and accurate as compared to using velocity from World, thus, trajectories generated in primitive executor may be slightly different from what was generated in AI, but this should be insignificant.
- Since almost all of the motion planning algorithm is performed in AI, `World`s do not need to be sent to the robot anymore, and `PrimitiveSet`s size can be reduced, reducing our overall bandwidth used over wifi. 

### Testing Done
- Added various unit tests to the newly added classes/methods
- Tested extensively in AI vs AI
- Tested on actual robots

### Length Justification and Key Files to Review
This PR changes the motion planning architecture from ground up, as a result, it's fairly thicc. Though many of the files can be skipped or include basic documentations and unit tests.

- ~2000 lines of added code in this PR is from the [AABB.cc](https://github.com/lohedges/aabbcc) (axis aligned bounding box tree datastructure) implementation which is from a separate source and was not implemented by me. As such, this section of the code can be ignored by the reviewer! It is also very likely that with the future revisions of the trajectory planner, this code is removed to better support dynamic obstacles.
- All tactics had to be updated to return new `Primitive` type. The tactic functionalities should be unchanged and all of the same features as before are offered.
- All code related to the previous motion planning architecture (HRVO and ENLSVG) has been removed, resulting in the large number of deletions. 

### Resolved Issues
Closes #2519
Closes #2655
Closes #2913
Closes #2894
Closes #2676
Closes #2629
Closes #2496

### CODE REVIEW

**Comment by akhilveeraghanta** [comment]
Curved paths are so cool! Good work :100:

**Review by williamckha — DISMISSED** [review]
Fantastic work 🚀
Read through most of the changes and LGTM

File: src/proto/parameters.proto [hunk 1]
```diff
@@ -126,9 +126,8 @@ message PassingConfig
 
 message GoalieTacticConfig
 {
-    // Control the final speed of the goalie to be able to intercept shots better
-    required double goalie_final_speed = 1
-        [default = 0.0, (bounds).min_double_value = 0.0, (bounds).max_double_value = 3.0];
+    reserved 1;
+
```
Comment (itsarune) [inline src/proto/parameters.proto:130]: How come we don't need this field with the trajectory planner?

File: src/proto/primitive.proto [hunk 2]
```diff
@@ -110,32 +95,54 @@ message PowerControl
...
+    // sub_destination. i.e. Trajectories are always from (0,0)
+    Point start_position = 1;
+    // Could create new PointF class (uses floats) or PointS (uses
+    // shorts to represent the milli-meter position, might have to
+    // signed)
+    Point destination                  = 2;
+    Vector initial_velocity            = 3;
+    MaxAllowedSpeedMode max_speed_mode = 4;
+    Point sub_destination              = 5;
+    // If connection_time is non-zero, then the trajectory path is
```
Comment (itsarune) [inline src/proto/primitive.proto]: Out of curiosity, could we have more that one sub trajectory? What would the problem/advantage be with it?

File: src/proto/parameters.proto (hunk 1 above)
Comment (nimazareian) [reply src/proto/parameters.proto:130]: The bang bang trajectories don't support having a non-zero final velocity. Plus, Im pretty sure this field wasn't used

File: src/proto/primitive.proto (hunk 2 above)
Comment (nimazareian) [reply src/proto/primitive.proto]: The current implementation of the trajectory planner only supports one sub-destination/trajectory. This is mainly to reduce the complexity of the search for the best trajectory path. Searching for two sub destinations that create the best trajectory path, as opposed to one sub destination will make the search exponentially harder with the current implementation. Additionally, for almost every case, one subdestination is enough to get around obstacles we face for the an initial short term of the trajectory (first couple of seconds), and trajectories generated later into the future will take into account farther obstacles.

But just to future proof it, and to make the code possibly slightly less jank with `connection_time_s = 0` representing no sub-destination, I've made the fields repeated so we can support more complex trajectories with multiple sub-destinations as well.

File: src/shared/2021_robot_constants.cpp [hunk 3]
```diff
@@ -25,9 +25,9 @@ RobotConstants_t create2021RobotConstants(void)
         .motor_max_acceleration_m_per_s_2 = 4.5f,
 
         // Robot's linear movement constants
-        .robot_max_speed_m_per_s          = 5.000f,
-        .robot_max_acceleration_m_per_s_2 = 3.5f,
-        .robot_max_deceleration_m_per_s_2 = 6.0f,
+        .robot_max_speed_m_per_s          = 3.000f,
```
Comment (akhilveeraghanta) [inline src/shared/2021_robot_constants.cpp:28]: why does this have to slow down? or is this more accurate and 5 was too fast/didn't make any sense?

File: src/software/time/time.h [hunk 4]
```diff
@@ -59,7 +59,6 @@ class Time
      */
     explicit Time(double time_seconds);
 
-   private:
```
Comment (akhilveeraghanta) [inline src/software/time/time.h:62]: unintentional? 

if intentational, probably should change to accessor

File: src/software/time/duration.h [hunk 5]
```diff
@@ -107,6 +107,24 @@ class Duration : public Time
...
+    Duration& operator+=(const Duration& duration);
+
+    /**
+     * Defines the subtraction compound assignment operator for Durations. Allows Duration
+     * to be subtracted from this Durations
+     *
+     * @param duration the Duration to subtract from this Duration
+     * @return Reference to the duration that was updated
+     */
+    Duration& operator-=(const Duration& duration);
```
Comment (akhilveeraghanta) [inline src/software/time/duration.h:126]: is a negative duration valid? what happens when u subtract Duration::from_secs(0) - Duration::from_secs(1)

File: src/software/networking/proto_udp_listener.hpp [hunk 6]
```diff
@@ -66,16 +66,6 @@ class ProtoUdpListener
     void handleDataReception(const boost::system::error_code& error,
```
Comment (akhilveeraghanta) [inline src/software/networking/proto_udp_listener.hpp:66]: nit: this change should be part of another PR

File: src/software/ai/navigator/trajectory/trajectory_planner.h [hunk 7]
```diff
@@ -0,0 +1,166 @@
...
+     * @param obstacles List of all obstacles
+     * @return A trajectory path with only a single trajectory + its cost
+     */
+    TrajectoryPathWithCost getDirectTrajectoryWithCost(
+        const Point &start, const Point &destination, const Vector &initial_velocity,
+        const KinematicConstraints &constraints, aabb::Tree &obstacle_tree,
+        const std::vector<ObstaclePtr> &obstacles);
+
+    /**
+     * Given a trajectory path, calculate its cost
```
Comment (akhilveeraghanta) [inline src/software/ai/navigator/trajectory/trajectory_planner.h:63]: nit: wrong comment

File: src/software/ai/navigator/trajectory/trajectory_planner.h [hunk 8]
```diff
@@ -0,0 +1,166 @@
...
+     * Calculate the cost of the given trajectory path with cost
+     *
+     * @param traj_with_cost A complete trajectory path with cost
+     * @return The cost of the trajectory
+     */
+    double calculateCost(const TrajectoryPathWithCost &traj_with_cost) const;
+
+    /**
+     * Get a single trajectory with cost that goes directly from the start to the
+     * destination.
```
Comment (akhilveeraghanta) [inline src/software/ai/navigator/trajectory/trajectory_planner.h:47]: Is this in a straight line ignoring obstacles?

File: src/software/ai/hl/stp/play/play.cpp [hunk 9]
```diff
@@ -2,6 +2,8 @@
 
 #include <munkres/munkres.h>
 
+#include <numeric>
```
Comment (itsarune) [inline src/software/ai/hl/stp/play/play.cpp]: nit: what is this `<numeric>` include for?

File: src/software/ai/hl/stp/tactic/crease_defender/crease_defender_fsm.cpp [hunk 10]
```diff
@@ -35,7 +35,7 @@ void CreaseDefenderFSM::blockThreat(
     auto block_threat_point = findBlockThreatPoint(
         event.common.world.field(), event.control_params.enemy_threat_origin,
         event.control_params.crease_defender_alignment,
-        robot_navigation_obstacle_config.robot_obstacle_inflation_factor());
+        robot_navigation_obstacle_config.robot_obstacle_inflation_factor() + 0.5);
```
Comment (itsarune) [inline src/software/ai/hl/stp/tactic/crease_defender/crease_defender_fsm.cpp:40]: why do we need an additional inflation factor here? why isn't `robot_obstacle_inflation_factor` good enough?

File: src/software/ai/hl/stp/tactic/tactic.h [hunk 11]
```diff
@@ -106,7 +105,7 @@ class Tactic
     std::optional<RobotId> last_execution_robot;
 
    private:
-    std::unique_ptr<TbotsProto::Primitive> primitive;
+    std::shared_ptr<Primitive> primitive;
```
Comment (itsarune) [inline src/software/ai/hl/stp/tactic/tactic.h:108]: why did we make the decision to use a shared_ptr instead of a unique_ptr again?

File: src/software/ai/navigator/trajectory/trajectory_planner.cpp [hunk 12]
```diff
@@ -0,0 +1,321 @@
...
+             connection_time += SUB_DESTINATION_STEP_INTERVAL_SEC)
+        {
+            // Branch off of a copy of the initial trajectory at connection_time
+            // to move towards the actual destination.
+            TrajectoryPath traj_path_to_dest = sub_trajectory.traj_path;
+            traj_path_to_dest.append(connection_time, destination, constraints);
+
+            // Return early for this sub destination if the trajectory can
+            // not have a lower cost than the best trajectory.
+            if (traj_path_to_dest.getTotalTime() >= best_traj_with_cost.cost)
```
Comment (itsarune) [inline src/software/ai/navigator/trajectory/trajectory_planner.cpp:103]: why are you comparing `getTotalTime()` to cost, are these comparable?

Comment (itsarune) [reply src/software/ai/navigator/trajectory/trajectory_planner.cpp:103]: I guess we're using time as the measure of `cost`

File: src/software/ai/navigator/trajectory/trajectory_planner.cpp [hunk 13]
```diff
@@ -0,0 +1,321 @@
...
+    total_cost += (first_collision_position - destination).length();
+
+    total_cost += std::max(
+        0.0, (MAX_FUTURE_COLLISION_CHECK_SEC - traj_with_cost.first_collision_time_s));
+
+    total_cost += 3 * traj_with_cost.collision_duration_front_s;
+
+    total_cost += 1 * traj_with_cost.collision_duration_back_s;
+
+    return total_cost;
```
Comment (itsarune) [inline src/software/ai/navigator/trajectory/trajectory_planner.cpp:245]: this may be something we want to spend some effort tuning -> this could be a good task for someone on the motion planning project

Comment (itsarune) [reply src/software/ai/navigator/trajectory/trajectory_planner.cpp:245]: is it worth having it configurable via dynamic params? It might be too specific to be useful as a dynamic param

File: src/software/ai/navigator/trajectory/trajectory_planner_test.cpp [hunk 14]
```diff
@@ -0,0 +1,109 @@
+#include "software/ai/navigator/trajectory/trajectory_planner.h"
+
+#include <gtest/gtest.h>
+
+#include "software/ai/navigator/obstacle/robot_navigation_obstacle_factory.h"
+#include "software/test_util/test_util.h"
+
+class TrajectoryPlannerTest : public testing::Test
```
Comment (itsarune) [inline src/software/ai/navigator/trajectory/trajectory_planner_test.cpp:8]: there would be a couple of tests that would be nice to see:

- avoiding the enemy half + centre circle
- avoiding the enemy half (no centre circle avoidance)
- a path ending in a blocked obstacle (eg. enemy half with enemy half constraint)

File: src/software/ai/hl/stp/tactic/crease_defender/crease_defender_fsm.cpp (hunk 10 above)
Comment (itsarune) [reply src/software/ai/hl/stp/tactic/crease_defender/crease_defender_fsm.cpp:40]: after talking to nima: it's so that the crease defenders are more stable on the enemy defense line

File: src/extlibs/AABB/README.md
Comment (itsarune) [inline src/extlibs/AABB/README.md:1]: you should consider pulling these as a bazel `git` dependency

File: src/software/ai/hl/stp/tactic/crease_defender/crease_defender_fsm.cpp (hunk 10 above)
Comment (akhilveeraghanta) [reply src/software/ai/hl/stp/tactic/crease_defender/crease_defender_fsm.cpp:40]: why not just inflate somemore?

File: src/software/ai/hl/stp/tactic/primitive.h [hunk 15]
```diff
@@ -0,0 +1,53 @@
...
+
+    /**
+     * Gets the estimated cost of the primitive
+     *
+     * @return estimated cost of the primitive
+     */
+    double getEstimatedPrimitiveCost() const
+    {
+        return estimated_cost;
+    }
```
Comment (akhilveeraghanta) [inline src/software/ai/hl/stp/tactic/primitive.h:49]: why provide a default prank implementation? should u make this pure virtual as well?

File: src/software/ai/hl/stp/play/play.cpp (hunk 9 above)
Comment (nimazareian) [reply src/software/ai/hl/stp/play/play.cpp]: Good call. it was extra and it's been removed

File: src/software/ai/navigator/trajectory/trajectory_planner.cpp (hunk 13 above)
Comment (nimazareian) [reply src/software/ai/navigator/trajectory/trajectory_planner.cpp:245]: I think the optimization of trajectories is still an open ended discussion, as there are possible alternative approaches that don't use cost at all. So not sure if the effort of tuning this is needed. Though it's a good idea to keep an eye out and see if it sometimes picks weird/unexpected trajectories, then we can look at those specific scenarios.

File: src/extlibs/AABB/README.md
Comment (nimazareian) [reply src/extlibs/AABB/README.md:1]: I totally forgot about this. I've updated it!

File: src/software/ai/navigator/trajectory/trajectory_planner_test.cpp (hunk 14 above)
Comment (nimazareian) [reply src/software/ai/navigator/trajectory/trajectory_planner_test.cpp:8]: I've added some sort of a test for the first two, but didn't add one for the last one since we currently don't have the implementation for moving destinations inside obstacles out of obstacles. This is something that is being tracked here #3105

**Review by itsarune — APPROVED** [review]
nice work!!

excited to see in my VM at 40-60 Hz!

**Review by williamckha — APPROVED** [review]
